# Packages used in this workflow

__BLAST__  
Altschul, S.F. et al. (1990) ‘Basic local alignment search tool’, Journal of molecular biology, 215(3), pp. 403–410. Available at: https://doi.org/10.1016/S0022-2836(05)80360-2.

__QUAST__  
Gurevich, A. et al. (2013) ‘QUAST: quality assessment tool for genome assemblies’, Bioinformatics , 29(8), pp. 1072–1075. Available at: https://doi.org/10.1093/bioinformatics/btt086.

__Snakemake__  
Köster, J. and Rahmann, S. (2012) ‘Snakemake—a scalable bioinformatics workflow engine’, Bioinformatics , 28(19), pp. 2520–2522. Available at: https://doi.org/10.1093/bioinformatics/bts480.

__Blobtools__  
Laetsch, D.R. and Blaxter, M.L. (2017) ‘BlobTools: Interrogation of genome assemblies’, F1000Research, 6(1287), p. 1287. Available at: https://doi.org/10.12688/f1000research.12232.1.

__Samtools__  
Li et al. (no date) ‘The Sequence alignment/map (SAM) format and SAMtools’, Bioinformatics  [Preprint]. Available at: http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.325.1516&rep=rep1&type=pdf.

__MUMmer__  
Marçais, G. et al. (2018) ‘MUMmer4: A fast and versatile genome alignment system’, PLoS computational biology, 14(1), p. e1005944. Available at: https://doi.org/10.1371/journal.pcbi.1005944.

__CEGMA__  
Parra, G., Bradnam, K. and Korf, I. (2007) ‘CEGMA: a pipeline to accurately annotate core genes in eukaryotic genomes’, Bioinformatics , 23(9), pp. 1061–1067. Available at: https://doi.org/10.1093/bioinformatics/btm071.

__Mosdepth__  
Pedersen, B. (no date) mosdepth: fast BAM/CRAM depth calculation for WGS, exome, or targeted sequencing. Github. Available at: https://github.com/brentp/mosdepth (Accessed: 21 October 2022).

__GenomeScope2__ __and__ __Smudgeplot__  
Ranallo-Benavidez, T.R., Jaron, K.S. and Schatz, M.C. (2020) ‘GenomeScope 2.0 and Smudgeplot for reference-free profiling of polyploid genomes’, Nature communications, 11(1), p. 1432. Available at: https://doi.org/10.1038/s41467-020-14998-3.

__Sniffles__  
Sedlazeck, F. (no date) Sniffles: Structural variation caller using third generation sequencing. Github. Available at: https://github.com/fritzsedlazeck/Sniffles (Accessed: 26 October 2021).

__BUSCO__  
Simão, F.A. et al. (2015) ‘BUSCO: assessing genome assembly and annotation completeness with single-copy orthologs’, Bioinformatics , 31(19), pp. 3210–3212. Available at: https://doi.org/10.1093/bioinformatics/btv351.